doc: improve description of module `exports`

sam-github opened the pull request:
##### Checklist

- [x] documentation is changed or added
- [x] commit message follows commit guidelines

##### Affected core subsystem(s)

doc

##### Description of change

   doc: improve description of module `exports`
    
    - Do not use word alias, it isn't well defined
    - Fix return value of require() example, which confusingly was not the
      exported API as it should have been.
    - Describe best practice in keeping exports and module.exports bound
      together.
    - Describe why exports exists
    - Remove reference to magic, which is also not well defined

Replace https://github.com/nodejs/node/pull/9552

Fix https://github.com/nodejs/node/pull/9552#issuecomment-260721718

iamchenxin commented:
What about `this`, i am not sure if i missed something. i can not find the definition of `this` in old doc.

sam-github commented:
Above PRs/issues address  specific comments you have made.

Do you have more, or can we close this PR?

iamchenxin commented:
Yes, seems `this` will be solved in 9623

iamchenxin commented:
And seems document this will make a break change. Cause of some third part tools set different values for `this`. I should check the `babel-plugin` again, to test if it is still break `this`.

iamchenxin commented:
Thanks for your patience again, open an issue in babel for asking `this`.
I gotta sleep, have a good day.

jasnell commented on doc/api/modules.md:
```diff
@@ -529,34 +529,43 @@ const x = require('./x');
-to `module.exports`. As with any variable, if you assign a new value to it, it
-is no longer bound to the previous value.
+The `exports` variable that is available within a module starts as if
+`module.exports` had been assigned to it before your module was evaluated.
```
Please avoid the use of informal pronouns such as `you`, `your` and `us` in the docs :-)

jasnell commented on doc/api/modules.md:
```diff
@@ -529,34 +529,43 @@ const x = require('./x');
+
+It allows a shortcut, so that `module.exports.f = ...` can be written more
+succinctly as `exports.f = ...`. However, be aware that like any variable, if
+you assign a new value to it, it is no longer bound to `module.exports`:
```
For example, this last sentence is better written as: "Note, however, assigning a new value to `exports` causes it to no longer be bound to `module.exports`."

sam-github commented:
Thanks @jasnell, I removed the colloquialisms, and also fixed another bug in the require pseudo-code.

Personally, I would remove the pseudo code in preference for the textual description and example, but maybe some other time, so it doesn't slow this PR.

jasnell commented on doc/api/modules.md:
```diff
@@ -529,34 +529,47 @@ const x = require('./x');
+succinctly as `exports.f = ...`. However, be aware that like any variable, if
+a new value is assigned to `exports`, it is no longer bound to `module.exports`:
+
+   module.exports.hello = true; // Exported from require of module
```
these should be wrapped in an explicit code block using the three backticks...

sam-github replied:
Done, thanks, sorry for the churn, I'll be up to speed on current conventions soon.

Qard commented:
LGTM

lpinca commented on doc/api/modules.md:
````diff
@@ -529,34 +529,51 @@ const x = require('./x');
+```
+
+When the `module.exports` property is being completely replaced by a new
+object, it is common to also reassign `exports`, for example:
````
Is this really widely used or a good practice?

Qard replied:
I wouldn't say it's universal, but I see it use fairly often as a defensive measure. I'd consider it a relatively good practice.

sam-github replied:
@lpinca: https://github.com/search?p=2&q=%22module.exports+%3D+exports%22&type=Code&utf8=%E2%9C%93

sam-github commented:
@jasnell can you confirm I made your requested changes?

sam-github commented:
@jasnell I don't want to merge while your review shows "requested changes" - I made the changes, can you confirm/LGTM?

sam-github commented:
Thanks!
